### DISCUSSION TITLE
New experimental asset graph layout algorithms

### DISCUSSION THREAD

**Discussion by salazarm** [dagster-io/dagster#17240] [Ideas]
We are experimenting with new and exponentially faster layout algorithms to dramatically improve the loading of asset graphs with 1,000+ assets. Two new feature flags have been added under user settings:
<img width="481" alt="Screenshot 2023-10-16 at 3 20 25 PM" src="https://github.com/dagster-io/dagster/assets/2286579/3e124789-d0a0-4190-8d6d-c1c4588dc876">

Below are the technical definitions. Network-simplex is the current algorithm which is iterative and is too slow for very large (1,000+ assets) graphs. We are considering changing to one of the other two options (tight-tree or longest-path) and would love your input into which you would prefer and why.
<img width="675" alt="Screenshot 2023-10-16 at 3 15 35 PM" src="https://github.com/dagster-io/dagster/assets/2286579/9c620a68-436c-42d6-a974-2117a79f487f">

The above definition is very technical but at a high level one of the main differences is that longest-path will align all of the leaf nodes at the bottom where as tight-tree aligns all of the root nodes at the top. Another difference is that when expanding/collapsing subgraphs, longest-path results in graphs that are not drastically different from their unexpanded/uncollapsed counterparts.

Please experiment with the options and let us know what you think!

(Note that you should only enable one of the flags at a time, otherwise tight-tree will take precedence over longest-path)

**Comment by maxwelljoslyn** [comment]
Here's a subset of my asset graph (with redacted asset and group names).

I'm using **longest-path** in this screenshot, and frankly it looks pretty bad. It's hard to read the arrows when they overlap like this.

<img width="1172" alt="Screen Shot 2023-11-07 at 14 15 42 PST" src="https://github.com/dagster-io/dagster/assets/11641081/58998007-c52f-43af-b3a6-65f5fb0ecedf">

I also tried **tight-tree**. That layout ended up almost the same as simplex, except that it moved some groups to different positions. But I didn't have any readability problems. So in the absence of any other information, **tight-tree** would definitely be my preference.

**Reply by maxwelljoslyn** [reply]
To put it another way, I think that the priority for graph drawing algorithms should be ensuring that arrows don't overlap. I don't really have a preference how that happens internally.

**Reply by maxwelljoslyn** [reply]
Since I'm on the subject of layout algorithms: I would find it useful to have layout use both horizontal and vertical space. One way to measure that would be a constraint that tries to minimize the area of an imaginary square with its upper left corner at the uppermost, leftmost asset, and its bottom right corner at the bottom-most, rightmost asset. That would help me see more stuff on screen without scrolling.
